client/my-sites/checkout/composite-checkout/wpcom/hooks/use-shopping-cart.ts

```diff
@@ -401,16 +404,24 @@ export function useShoppingCart(
 	productsToAdd: RequestCartProduct[] | null,
 	couponToAdd: string | null,
 	setCart: ( arg0: string, arg1: RequestCart ) => Promise< ResponseCart >,
-	getCart: ( arg0: string ) => Promise< ResponseCart >,
+	getCart: ( arg0: string, arg1: ResponseCart | null ) => Promise< ResponseCart >,
 	showAddCouponSuccessMessage: ( arg0: string ) => void,
+	cartRawResponse: ResponseCart,
 	onEvent?: ( arg0: ReactStandardAction ) => void
 ): ShoppingCartManager {
 	const cartKeyString: string = cartKey || 'no-site';
 	const setServerCart = useCallback( ( cartParam ) => setCart( cartKeyString, cartParam ), [
 		cartKeyString,
 		setCart,
 	] );
-	const getServerCart = useCallback( () => getCart( cartKeyString ), [ cartKeyString, getCart ] );
+
+	const cartResponseConverted = isEmpty( cartRawResponse )
+		? null
+		: convertRawResponseCartToResponseCart( cartRawResponse );
+	const getServerCart = useCallback( () => getCart( cartKeyString, cartResponseConverted ), [
```

Comment:
Agreed 👍 .I've pushed in a change that makes these suggested changes.

Reply:
Hm... interesting! The issue, as I understand it, is that the fundamental behavior of this hook is being changed. 

Previously, the hook would wait for the `CartStore` to be done loading (using the `canInitializeCart` boolean passed to the hook), and then fetch the cart from the server. That test verifies that the hook will wait for the `CartStore` to be finished with any operations before it attempts to fetch the cart (to protect against race conditions).

With this PR, the hook will actually never fetch the cart itself if the cart already exists. (In fact, it occurs to me that we could remove the `getServerCart` and that whole set of functions entirely because `cartRawResponse` will always be set.) The result is that the failing test (and many of the other tests) no longer are testing the same things. This particular test is failing because the cart object being passed in via the test isn't a full `ResponseCart`.

This modification to the hook is bigger than I thought; 😁  I'd like to experiment with some refactoring of my own and see if I can alter the hook to work well for both of our purposes. I'll reply to this soon with my results. I'll make my own branch so that we can be sure it will work for checkout without userless but I'll probably take many of your changes from here.

Reply:
I think I actually found an intermediate solution that will only affect userless checkout and I think will still work for both of our cases (but please tell me if it works for you). I pushed a change to this branch; I hope that's ok!

Basically, `useShoppingCart` accepts a `getCart` function, which it uses to fetch the cart. Originally in this PR you were overloading that function to handle both use cases and then passing `useShoppingCart` the `CartStore`'s data so it could decide which of the two versions to use. With my change, we instead use the already-existing `getCart` override prop built into `CompositeCheckout` to override fetching the cart from the server if we are using userless checkout. 

The only downside might be that this will only be used for the initial fetch; after that we always use the cart data as returned from the shopping-cart endpoint. If this is an issue for you, we may want to also override `setCart`.

Reply:
This works well, thanks. I pushed a small change for `Promise.new` to change to `Promise.resolve`.

> The only downside might be that this will only be used for the initial fetch; after that we always use the cart data as returned from the shopping-cart endpoint.

I think this should be okay. In the logged out cart, since the cart items are local to that browser and device. 

>  I pushed a change to this branch; I hope that's ok!
Of course, please feel free :)